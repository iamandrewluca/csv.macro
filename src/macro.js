// @flow
import fs from 'fs';
import { createMacro } from 'babel-plugin-macros';
import parse from 'csv-parse/lib/sync';
import resolveImportPath from './resolveImportPath';

function csvMacro({
  references: { loader = [] },
  state: {
    file: {
      opts: { filename },
    },
  },
}: {
  references: { loader: Array<any> },
  state: { file: { opts: { filename: string } } },
  babel: { types: Object },
}): void {
  // import { loader } from 'csv.macro'
  loader.forEach(referencePath => {
    referencePath.parentPath.node.arguments.forEach(({ value }) => {
      const absolutePath = resolveImportPath({ filename, relativePath: value });
      const csvContent = fs.readFileSync(absolutePath, 'utf8');
      const json = parse(csvContent, { columns: true, skip_empty_lines: true });
      referencePath.parentPath.replaceWithSourceString(JSON.stringify(json));
    });
  });
}

export default createMacro(csvMacro);
