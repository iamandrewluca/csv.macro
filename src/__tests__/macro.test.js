// @flow

import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';
import fs from 'fs';
import path from 'path';

expect.addSnapshotSerializer({
  print(val) {
    return val.replace(/..\/macro/, 'csv.macro');
  },
  test(val) {
    return typeof val === 'string';
  },
});

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    filename: __filename,
  },
  tests: {
    '[loader] with data': {
      error: false,
      code: `
        import { loader } from '../macro';
        const jsonData = loader('./fixtures/data.csv');
      `,
    },
    '[loader] with absolute path': {
      error: false,
      code: `
        import { loader } from '../macro';
        const jsonData = loader('src/__tests__/fixtures/data.csv');
      `,
    },
    '[loader] with absolute path and NODE_PATH': {
      error: false,
      code: `
        import { loader } from '../macro';
        const jsonData = loader('__tests__/fixtures/data.csv');
      `,
      setup: () => {
        process.env.NODE_PATH = 'src/';
      },
    },
    '[loader] with absolute path and jsconfig include property': {
      error: false,
      code: `
        import { loader } from '../macro';
        const jsonData = loader('__tests__/fixtures/data.csv');
      `,
      setup: () => {
        fs.symlinkSync(
          'src/__tests__/fixtures/jsconfig.json',
          path.resolve(fs.realpathSync(process.cwd()), 'jsconfig.json'),
          'file',
        );
      },
      teardown: () => {
        fs.unlinkSync(
          path.resolve(fs.realpathSync(process.cwd()), 'jsconfig.json'),
        );
      },
    },
    '[loader] should work with relative path': {
      error: false,
      code: `
        import { loader } from '../macro';
        const jsonData = loader('../__tests__/fixtures/data.csv');
      `,
    },
  },
});
