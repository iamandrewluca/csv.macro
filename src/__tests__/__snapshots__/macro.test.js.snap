// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros [loader] should work with relative path: [loader] should work with relative path 1`] = `

import { loader } from 'csv.macro';
const jsonData = loader('../__tests__/fixtures/data.csv');

      ↓ ↓ ↓ ↓ ↓ ↓

const jsonData = [{
  "id": "1",
  "username": "andrew",
  "email": "andrew@example.com"
}, {
  "id": "2",
  "username": "luca",
  "email": "luca@example.com"
}];

`;

exports[`macros [loader] with absolute path and NODE_PATH: [loader] with absolute path and NODE_PATH 1`] = `

import { loader } from 'csv.macro';
const jsonData = loader('__tests__/fixtures/data.csv');

      ↓ ↓ ↓ ↓ ↓ ↓

const jsonData = [{
  "id": "1",
  "username": "andrew",
  "email": "andrew@example.com"
}, {
  "id": "2",
  "username": "luca",
  "email": "luca@example.com"
}];

`;

exports[`macros [loader] with absolute path and jsconfig include property: [loader] with absolute path and jsconfig include property 1`] = `

import { loader } from 'csv.macro';
const jsonData = loader('__tests__/fixtures/data.csv');

      ↓ ↓ ↓ ↓ ↓ ↓

const jsonData = [{
  "id": "1",
  "username": "andrew",
  "email": "andrew@example.com"
}, {
  "id": "2",
  "username": "luca",
  "email": "luca@example.com"
}];

`;

exports[`macros [loader] with absolute path: [loader] with absolute path 1`] = `

import { loader } from 'csv.macro';
const jsonData = loader('src/__tests__/fixtures/data.csv');

      ↓ ↓ ↓ ↓ ↓ ↓

const jsonData = [{
  "id": "1",
  "username": "andrew",
  "email": "andrew@example.com"
}, {
  "id": "2",
  "username": "luca",
  "email": "luca@example.com"
}];

`;

exports[`macros [loader] with data: [loader] with data 1`] = `

import { loader } from 'csv.macro';
const jsonData = loader('./fixtures/data.csv');

      ↓ ↓ ↓ ↓ ↓ ↓

const jsonData = [{
  "id": "1",
  "username": "andrew",
  "email": "andrew@example.com"
}, {
  "id": "2",
  "username": "luca",
  "email": "luca@example.com"
}];

`;
